d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -212,63 +211,87 @@ private void readyHandler()
     }
   }
 
-  private void handleD2NodeResponse(DiscoveryResponseData data)
+  @VisibleForTesting
+  void handleD2NodeResponse(DiscoveryResponseData data)
   {
     Map<String, NodeUpdate> updates = new HashMap<>();
     List<String> errors = new ArrayList<>();
 
-    for (Resource resource: data.getResourcesList())
+    for (Resource resource : data.getResourcesList())
     {
       String resourceName = resource.getName();
       try
       {
         XdsD2.Node d2Node = resource.getResource().unpack(XdsD2.Node.class);
+        if (d2Node != null && d2Node.getData().isEmpty())
+        {
+          _log.warn("Received a Node response with no data, resource is : {}", resourceName);
+        }
         updates.put(resourceName, new NodeUpdate(resource.getVersion(), d2Node));
-      } catch (InvalidProtocolBufferException e)
+      }
+      catch (InvalidProtocolBufferException e)
       {
         _log.warn("Failed to unpack Node response", e);
         errors.add("Failed to unpack Node response");
+        // Assume that the resource doesn't exist if it cannot be deserialized instead of simply ignoring it. This way
+        // any call waiting on the response can be satisfied instead of timing out.
+        updates.put(resourceName, new NodeUpdate(resource.getVersion(), null));
       }
     }
-
-    handleResourceUpdate(updates, data.getResourceType(), data.getNonce(), errors);
+    sendAckOrNack(data.getResourceType(), data.getNonce(), errors);
+    handleResourceUpdate(updates, data.getResourceType());
+    handleResourceRemoval(data.getRemovedResources(), data.getResourceType());
   }
 
-  private void handleD2URIMapResponse(DiscoveryResponseData data)
+  @VisibleForTesting
+  void handleD2URIMapResponse(DiscoveryResponseData data)
   {
     Map<String, D2URIMapUpdate> updates = new HashMap<>();
     List<String> errors = new ArrayList<>();
 
-    for (Resource resource: data.getResourcesList())
+    for (Resource resource : data.getResourcesList())
     {
       String resourceName = resource.getName();
       try
       {
         XdsD2.D2URIMap uriMap = resource.getResource().unpack(XdsD2.D2URIMap.class);
         Map<String, XdsD2.D2URI> nodeData = uriMap.getUrisMap();
+        if (nodeData.isEmpty())
+        {
+          _log.warn("Received a D2URIMap response with no data, resource is : {}", resourceName);
```

Comment:
Better let it be info. App owners wouldn't know how to interpret it anyway, and we might get some unneeded escalations.

Reply:
Actually, we should not downlevel the log from `warn` to `info` as this is a truly warn things that we need to catch. In normal scenario, there should not be empty data sent from `indis`